libsolidity/formal/SMTChecker.h

```diff
@@ -161,12 +161,14 @@ class SMTChecker: private ASTConstVisitor
 	/// Add to the solver: the given expression implied by the current path conditions
 	void addPathImpliedExpression(smt::Expression const& _e);
 
+	/// Clears the local variables of a function.
+	void clearLocalVariables();
+
 	std::shared_ptr<smt::SolverInterface> m_interface;
 	std::shared_ptr<VariableUsage> m_variableUsage;
 	bool m_loopExecutionHappened = false;
 	std::map<Expression const*, smt::Expression> m_expressions;
 	std::map<Declaration const*, SSAVariable> m_variables;
```

Comment:
Do we actually need this to be a `Declaration`? Why can't we use `VariableDeclaration` here?

Reply:
I thought the same yesterday so I checked why: the reason is that we sometimes use `referencedDeclaration` which is a `Declaration const*`.

Reply:
Sure, but does it makes sense to perform the cast at the point where the map is filled instead of every time we access it? Do we intend to put something inside that is not a VariableDeclaration?

Reply:
Only `VariableDeclaration` goes in.
If we declare it as `VariableDeclaration` we have to cast when accessing for encoding, like in `knownVariable`, `newValue`, `setZeroValue` and `setUnknown`.
If we declare it as `Declaration` we have to cast when reseting local and state variables.
I'm fine with either, but I think the second case (current) performs fewer casts.

Reply:
We might want to change the argument of all those others, too. Does it make sense to call `knownVariable`, ... with something that is not a VariableDeclaration?

But yeah, let's do this in a different PR if at all.

Reply:
We can change those too, but any variable reference will require a cast anyway on `referencedDeclaration`.

Reply:
Let me check this better first. It's also not urgent, so we can think about it a bit.